src/platform/Linux/CHIPBluezHelper.cpp

```diff
@@ -1710,6 +1667,101 @@ CHIP_ERROR InitBluezBleLayer(bool aIsCentral, char * apBleAddr, BLEAdvConfig & a
     return err;
 }
 
+// StartDiscovery callbacks
+
+using DiscoveryTaskArg = std::pair<BluezEndpoint *, BluezDiscoveryRequest>;
+
+void StartDiscoveryDone(GObject * aObject, GAsyncResult * aResult, gpointer apEndpoint)
+{
+    BluezAdapter1 * adapter = BLUEZ_ADAPTER1(aObject);
+    GError * error          = nullptr;
+    gboolean success        = bluez_adapter1_call_start_discovery_finish(adapter, aResult, &error);
+
+    VerifyOrExit(success == TRUE, ChipLogProgress(DeviceLayer, "FAIL: StartDiscovery : %s", error->message));
+    ChipLogProgress(DeviceLayer, "StartDiscovery complete");
+
+exit:
+    if (error != nullptr)
+        g_error_free(error);
+}
+
+static gboolean StartDiscoveryImpl(void * apDiscoveryTaskArg)
+{
+    DiscoveryTaskArg * taskArg = static_cast<DiscoveryTaskArg *>(apDiscoveryTaskArg);
+    BluezEndpoint * endpoint;
+
+    VerifyOrExit(taskArg != nullptr, ChipLogProgress(DeviceLayer, "taskArg is NULL in %s", __func__));
+    endpoint = taskArg->first;
+
+    VerifyOrExit(endpoint != nullptr, ChipLogProgress(DeviceLayer, "endpoint is NULL in %s", __func__));
+    VerifyOrExit(endpoint->mpAdapter != nullptr, ChipLogProgress(DeviceLayer, "mpAdapter is NULL in %s", __func__));
+
+    endpoint->mDiscoveryRequest = taskArg->second;
+    delete taskArg;
```

Comment:
Hmm.. can we move it under `exit`? Or it's on purpose to not delete it if line 1696 and line 1697 fails?

```
if (taskArg)
   delete taskArg;
```

Reply:
Right, thanks!